src/renderer/atlas/AtlasEngine.h

```diff
@@ -0,0 +1,748 @@
+    constexpr bool operator==(const type& rhs) const noexcept  \
+    {                                                          \
+        return __builtin_memcmp(this, &rhs, sizeof(rhs)) == 0; \
+    }                                                          \
+                                                               \
+    constexpr bool operator!=(const type& rhs) const noexcept  \
+    {                                                          \
+        return __builtin_memcmp(this, &rhs, sizeof(rhs)) != 0; \
+    }
+
+#define ATLAS_FLAG_OPS(type, underlying)                                                                                                                    \
+    friend constexpr type operator~(type v) noexcept { return static_cast<type>(~static_cast<underlying>(v)); }                                             \
+    friend constexpr type operator|(type lhs, type rhs) noexcept { return static_cast<type>(static_cast<underlying>(lhs) | static_cast<underlying>(rhs)); } \
+    friend constexpr type operator&(type lhs, type rhs) noexcept { return static_cast<type>(static_cast<underlying>(lhs) & static_cast<underlying>(rhs)); } \
+    friend constexpr void operator|=(type& lhs, type rhs) noexcept { lhs = lhs | rhs; }                                                                     \
+    friend constexpr void operator&=(type& lhs, type rhs) noexcept { lhs = lhs & rhs; }
+
+        template<typename T>
+        struct vec2
+        {
+            T x{};
+            T y{};
+
+            ATLAS_POD_OPS(vec2)
+
+            constexpr vec2 operator/(const vec2& rhs) noexcept
+            {
+                assert(rhs.x != 0 && rhs.y != 0);
+                return { gsl::narrow_cast<T>(x / rhs.x), gsl::narrow_cast<T>(y / rhs.y) };
+            }
+        };
+
+        template<typename T>
+        struct vec4
+        {
+            T x{};
+            T y{};
+            T z{};
+            T w{};
+
+            ATLAS_POD_OPS(vec4)
+        };
+
+        template<typename T>
+        struct rect
+        {
+            T left{};
+            T top{};
+            T right{};
+            T bottom{};
+
+            ATLAS_POD_OPS(rect)
+
+            constexpr bool non_empty() noexcept
+            {
+                return (left < right) & (top < bottom);
+            }
+        };
+
+        using u8 = uint8_t;
+
+        using u16 = uint16_t;
+        using u16x2 = vec2<u16>;
+        using u16r = rect<u16>;
+
+        using i16 = int16_t;
+
+        using u32 = uint32_t;
+        using u32x2 = vec2<u32>;
+
+        using i32 = int32_t;
+
+        using f32 = float;
+        using f32x2 = vec2<f32>;
+        using f32x4 = vec4<f32>;
+
+        struct TextAnalyzerResult
+        {
+            u32 textPosition = 0;
+            u32 textLength = 0;
+
+            // These 2 fields represent DWRITE_SCRIPT_ANALYSIS.
+            // Not using DWRITE_SCRIPT_ANALYSIS drops the struct size from 20 down to 12 bytes.

```

Comment:
It's just to fit more data into the same amount of memory.
The reason I did this is because that `DWRITE_SCRIPT_SHAPES` field (here as `u8 shapes`) would normally consume an extra 8 bytes of struct size despite literally only being 1 bit large (0 or 1).  That kinda pissed me off a bit lol. 😐